compiler/rustc_metadata/src/locator.rs

```diff
@@ -783,14 +789,67 @@ fn get_metadata_section<'p>(
     filename: &'p Path,
     loader: &dyn MetadataLoader,
     cfg_version: &'static str,
+    crate_kind: ExternCrateKind,
+    crate_name: Option<Symbol>,
 ) -> Result<MetadataBlob, MetadataError<'p>> {
     if !filename.exists() {
         return Err(MetadataError::NotPresent(filename));
     }
+
+    // FIXME: Allow it for statically linked dependencies as well.
+    if crate_kind == ExternCrateKind::Stable && flavor != CrateFlavor::Dylib {
+        return Err(MetadataError::LoadFailure(
+            "`extern dyn` annotation is only avaible for dynamic dependencies".to_string(),
+        ));
+    }
+
     let raw_bytes = match flavor {
         CrateFlavor::Rlib => {
             loader.get_rlib_metadata(target, filename).map_err(MetadataError::LoadFailure)?
         }
+        CrateFlavor::Dylib if crate_kind == ExternCrateKind::Stable => {
+            let compiler = std::env::current_exe().map_err(|_err| {
+                MetadataError::LoadFailure(
+                    "couldn't obtain current compiler binary when loading `extern dyn` dependency"
+                        .to_string(),
+                )
+            })?;
+
+            let mut interface = PathBuf::from(filename);
+            interface.set_extension("rs");
+            debug!("compiling {}", interface.display());
+
+            let interface_dir = interface.parent().unwrap();
+            let crate_name = crate_name.unwrap();
+
+            let res = std::process::Command::new(compiler)
```

Comment:
This will need to be done either within the current compiler session or as a separate compiler session in the same process. As implemented this will break rustdoc and any other custom driver that has a different cli from rustc itself.